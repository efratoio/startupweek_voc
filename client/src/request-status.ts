export enum RequestStatus {
    NONE = "none",
    IN_PROGRESS = "in_progress",
    FAIL = "fail",
    SUCCESS = "success",
}
