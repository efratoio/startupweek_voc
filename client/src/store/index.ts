export * from "./api";
export * from "./signup";
export * from "./transactions";
