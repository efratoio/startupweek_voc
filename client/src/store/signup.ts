import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { ApiStore } from ".";
import { History } from "history";
import { component, inject } from "tsdi";

@component
export class SignupStore {
    @inject private api: ApiStore;

    @observable public signupResult: Boolean;

    @bind @action
    public async doSignup(email: string, password: string) {
        const body = { email, password };
        const response = await this.api.call(
            "doSignup",
            async () => {
                // TODO: Perform request here!
                return {};
            },
            true,
        );
        if (response) {
            await this.api.doLogin(email, password);
        }
    }
}
