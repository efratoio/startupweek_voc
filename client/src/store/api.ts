import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import * as isomorphicFetch from "isomorphic-fetch";
import { RequestStatus } from "../request-status";
import { History } from "history";
import { routeDashboard } from "../routing";
import { component, inject, initialize } from "tsdi";
import { snake, camel } from "change-case";

declare var baseUrl: string;
const prefixedBaseUrl = `${window.location.protocol}//${baseUrl}`;

const softwareVersion = 2;
const localStorageIdentifier = "accounting-software-login";

interface LocalStorageApi {
    storageVersion: number;
    date: string;
    authToken: string;
    userId: string;
}

interface ApiError {
    message: string;
}

interface RequestMetaInfo {
    error?: ApiError;
    requestStatus: RequestStatus;
}

export function convertCase(obj: any, conversion: (input: string) => string = snake): any {
    if (Array.isArray(obj)) {
        return obj.map(value => convertCase(value, conversion));
    }
    if (typeof obj === "object" && obj !== null) { // tslint:disable-line
        return Object.keys(obj).reduce((result, key) => {
            const value = obj[key];
            const newKey = conversion(key);
            result[newKey] = convertCase(value);
            return result;
        }, {});
    }
    return obj;
}

@component
export class ApiStore {
    @inject private browserHistory: History;

    @observable public authToken: string;
    @observable public userId: string;
    @observable public requests: Map<string, RequestMetaInfo> = new Map();
    @observable public errors: ApiError[] = [];

    @initialize
    protected init() {
        this.load();
    }

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined" && typeof this.userId !== "undefined";
    }

    @bind
    private wrappedFetch(url: string, options: any) {
        const headers = new Headers();
        if (options.headers) {
            Object.keys(options.headers).forEach(key => headers.append(key, options.headers[key]));
        }
        if (typeof options.body === "string") {
            options.body = JSON.stringify(convertCase(JSON.parse(options.body), snake));
        }
        if (this.loggedIn) {
            headers.append("Authorization", `Bearer ${this.userId}:${this.authToken}`);
        }
        return isomorphicFetch(url, { ...options, headers });
    }

    @bind @action
    public async call<T>(identifier: string, request: () => Promise<T>, supressErrors = false): Promise<T> {
        try {
            this.requests.set(identifier,{ requestStatus: RequestStatus.IN_PROGRESS });
            const response = await request();
            this.requests.set(identifier, { requestStatus: RequestStatus.SUCCESS });
            return convertCase(response, camel) as T;
        } catch (error) {
            this.requests.set(identifier, { requestStatus: RequestStatus.FAIL, error });
            if (!supressErrors) {
                this.errors.push(error);
            }
            return undefined;
        }
    }

    @bind @action
    public async doLogin(email: string, password: string) {
        const body = { email, password };
        const response = await this.call(
            "doLogin",
            async () => {
                // TODO: Perform request here!
                return {
                    id: "some-user-id",
                    token: "some-auth-token",
                };
            },
            true,
        );
        if (response) {
            const { id, token } = response;
            this.authToken = token;
            this.userId = id;
            this.save();
            this.browserHistory.replace(routeDashboard());
        }
    }

    @bind @action
    public logout() {
        this.clearStorage();
        this.authToken = undefined;
        this.userId = undefined;
    }

    @bind
    private save() {
        const deserialized: LocalStorageApi = {
            storageVersion: softwareVersion,
            date: new Date().toString(),
            authToken: this.authToken,
            userId: this.userId,
        };
        const serialized = JSON.stringify(deserialized);
        localStorage.setItem(localStorageIdentifier, serialized);
    }

    @bind
    private clearStorage() {
        localStorage.removeItem(localStorageIdentifier);
    }

    @bind
    private load() {
        const serialized = localStorage.getItem(localStorageIdentifier);
        if (serialized === null) { // tslint:disable-line
            return;
        }
        let deserialized: LocalStorageApi;
        try {
            deserialized = JSON.parse(serialized);
        } catch (err) {
            this.clearStorage();
            return;
        }
        if (deserialized.storageVersion !== softwareVersion) {
            this.clearStorage();
            return;
        }
        this.authToken = deserialized.authToken;
        this.userId = deserialized.userId;
        this.browserHistory.replace(routeDashboard());
    }

    public getRequestStatus(identifier: string): RequestStatus {
        const request = this.requests.get(identifier);
        if (typeof request === "undefined") {
            return RequestStatus.NONE;
        }
        return request.requestStatus;
    }

    public getError(identifier: string): ApiError {
        const request = this.requests.get(identifier);
        if (typeof request === "undefined") {
            return undefined;
        }
        return request.error;
    }

    @bind @action
    public doDismiss() {
        this.errors.pop();
    }

    @computed
    public get latestError() {
        return this.errors[this.errors.length - 1];
    }
}
