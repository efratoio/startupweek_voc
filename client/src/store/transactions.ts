import { component, inject, initialize } from "tsdi";
import { TransactionsResult, transactions, processSpeech, save } from "api";
import { ApiStore } from ".";
import { observable, action, computed } from "mobx";
import { Transaction } from "types";
import { bind } from "bind-decorator";

const now = new Date();

@component
export class TransactionsStore {
    @inject private api: ApiStore;

    @observable public transactions: Transaction[] = [];
    @observable public unsavedTransaction: Transaction;

    @computed public get byCategory(): { [category: string]: number } {
        return this.transactions.reduce((result, transaction) => {
            const { category, amount } = transaction;
            const current = result[category] || 0;
            result[transaction.category] = current + amount;
            return result;
        }, {});
    }

    @action @initialize
    public async doLoadTransactions() {
        const response = await this.api.call(
            "doLoadTransactions",
            () => transactions(this.api.authToken),
        );
        if (response) {
            this.transactions = response.map((item) => ({
                amount: parseFloat(item[2]),
                currency: item[3],
                category: item[1]
            }));
        }
    }

    @action
    public async doGenerateTransactionFromString(data: string) {
        const response = await this.api.call(
            "doGenerateTransactionFromString",
            () => processSpeech(data)
        );

        if (response) {
            this.unsavedTransaction = response;
        }
    }

    @action
    public async doSave(category: string, amount: number) {
        const body = { category, amount };
        const response = await this.api.call(
            "doSubmitItem",
            async () => save(this.api.authToken, category, amount, "$"),
            true,
        );
        if (response) {
            this.transactions.push(response);
        }
    }
}
