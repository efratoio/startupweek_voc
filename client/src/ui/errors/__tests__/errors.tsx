import * as React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { RequestStatus } from "request-status";
import { Errors } from "..";
import { Provider } from "mobx-react";
import { ApiStore } from "store";

test("\`Errors\` is rendered correctly with a network error", () => {
    tsdi.get(ApiStore).errors.push({ message: "some.message" });
    const mounted = mount(<Errors />);
    expect(toJson(mounted)).toMatchSnapshot();
});

test("\`Errors\` is renders nothing without an error", () => {
    const mounted = mount(<Errors />);
    expect(toJson(mounted)).toMatchSnapshot();
});
