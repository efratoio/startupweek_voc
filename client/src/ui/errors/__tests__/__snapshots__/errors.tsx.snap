// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Errors\` is rendered correctly with a network error 1`] = `
<Errors>
  <ErrorComponent
    message="some.message"
    title="An Error Occured"
  >
    <div>
      <h1>
        An Error Occured
      </h1>
      <p>
        some.message
      </p>
    </div>
  </ErrorComponent>
</Errors>
`;

exports[`\`Errors\` is renders nothing without an error 1`] = `<Errors />`;
